### PR TITLE
Guest accounts feature

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Guest accounts feature

### CODE REVIEW

**Comment by ogi-m** [comment]
### Invite flow issues: 

**1. "Copy Link" button doesn't work**
- Go to Invite People > Invite Members 
- Click "Copy Link" button next to the invite link
Observed: Text isn't copied and there's no visual confirmation that the button was clicked
Expected: Text is copied and there's visual confirmation that the button was clicked

**2. Blank screen when inviting new Members/Guests** 
1. Go to Invite People > Invite Members (same issue happens on Invite Guests)
2. Enter an email that doesn't exist in the system and click "Invite Members"
Observed: Screen goes blank (see screenshot for console errors)
Expected: Confirmation screen is shown saying an invite was sent to the email address

Screenshot (167)

**3. Invite modal can't be scrolled**
- Go to Invite People > Invite Guests
- Click "Set a custom message" to expand the "Custom message" field
Observed: "Invite Guests" button is pushed out out of view with no way to scroll it into view
Expected: User can scroll the modal

**Note**: 1080p display at 125% UI scaling


**4. Existing Guests are not added to the team**
- Create a team and go to Invite People > Invite Guests
- Select '@guest' user who is not a guest on the current team
- Select a channel for the guest user to join and click "Invite Guests"
- Success message is shown saying that the guest was added to the team
- Close the modal and view the members of the channel (and/or the team) the guest was added to
Observed: Guest user was not added to the channel or the team
Expected: The user is added to the team and to the specified channel

**Comment by jespino** [comment]
@ogi-m 
1. I can reproduce it locally
2. I fixed that here: https://github.com/mattermost/mattermost-webapp/pull/3102
3. This is that way by design I think <- @michaelgamble what do you think about the point 3?
4. Clearly something is wrong there, I have to check it properly (in some cases work, but not in all cases).
5. I'll check it
6. I'll check it
7. The current mattermost email validation is just that, we can be more strict about it, but have other implications.
8. @michaelgamble which is the good one here?
9. I checked that with @michaelgamble and I think the current styling is the correct one. Can you confirm that @michaelgamble?
10. I'll fix that
11. Yes, that is a know issue, I have to fix that too.
12. I don't agree with that, when you are inviting guest if you put regular users, the most important failure is that you can't add regular users as guests. @michaelgamble what do you think about it?
13. I'll fix that

**Comment by ogi-m** [comment]
@jespino I've updated issues 8, 9 and 12 with links to the design for reference.
> 1. I can't reproduce it locally

This one is a blocker - I can't create any guest users (And I can't reset the password for @guest).

**Comment by michaelgamble** [comment]
@jespino and @ogi-m 
3. I would expect the whole page to scroll with the exception of the "back" and "close" buttons would stay fixed on the page while scrolling. (scroll bar would be placed full height of screen right side)
8. Use the titles in the invision design as those were the ones that went through final reviews. "Invite People"
9. @ogi-m can you reach out to me on this one, I'm wondering if its the difference of which platform you are viewing in, but because I'm not sure i wouldn't mind doing a quick zoom to take a look.
12. The intention here is, the invite as a guest failed, however they most likely follow up action is "oh shoot I want to add them to the channel anyways" so thats why out of a few rounds of review and discussion we added that use case to the design. Lets stick with whats in invision on this one for now.

**Comment by michaelgamble** [comment]
@jespino 
9. reviewed with @ogi-m seems the weights are off here, it may have to do with the fact that on the previous screen its a mix of font-weight 300 and 400.. but on this screen it should be 'normal' which calculates to 400

Also I noticed on @ogi-m screenshots another issue of the input field heights seem to be wrong in chrome in windows - a discrepancy, we'll have to do a UX review pass in windows and see if anything else is out of place too.

**Comment by ogi-m** [comment]
Two more issues with the invite flow:

14. "Details'  text overlaps with a longer email
- Go to Invite People > Invite People
- Enter a longer email and click "Invite Members"
- Note the email and the "Details" text in the table on the summary page
Observed: Email and text overlap (see attached screenshot below)
Expected: No overlap

15. Wrong green color in dark mode
- Go to Account Settings > Display > Theme and select Mattermost Dark
- Go to Invite People > Invite People
- Enter an email and click "Invite Members"
- Note the green color in the table header
Observed: Different green from [the one in invision design](https://projects.invisionapp.com/d/main/default/?origin=v7#/console/17561763/364046127/preview)
Expected: Green matches invision
Screenshot (172)

**Comment by srkgupta** [comment]
Hi @jespino 

Observed an issue while testing the Guest Accounts feature. Can you please check and fix the same. 

**Whitelisted Domains for Guest Accounts does not works**
Steps to reproduce this issue:
- Navigate to System Console -> Guest Access Page. 
- Enter the Whitelisted Guest domains as "company.com"
- Now trigger the Guest Invitation process from the main screen and try to signup with an email belonging to a different domain other than the one provided in the whitelisted section
- User is allowed to create an account and login as Guest without any restriction.

**Comment by srkgupta** [comment]
Hi @jespino, 

Some other issues I observed during my initial testing:

1) When Guest account is turned off from the System Console, the option to Invite users as Guest is still shown in the main screen. Only when user clicks on it and proceeds to the next step, a message is displayed indicating that the Guest account featured needs to be turned on. 
2) Similarly if license is not uploaded on an instance, the option to invite users as Guest is still shown in the main screen. Only when user clicks on it and proceeds to the next step, a message is displayed indicating that the License needs to be uploaded in order to use the Guest Accounts feature. This might be problematic to GA instances without Enterprise licenses. We generally turn off the features which are not available, rather than showing an error message.

**Comment by wiersgallak** [comment]
@jespino Can we please also make sure this feature is labelled as "(Beta)" on the top of the Guest Access page and the left hand side listing in the Console?

**Comment by srkgupta** [comment]
Hi @jespino, 

Few other issues observed:

**1) Unable to Enforce MFA only for Guest Users**
	- Login to System Console and ensure MFA is enabled and Enforce MFA is set to false. 
	- Navigate to Guest Access page and check if Admin is allowed to Enforce MFA for Guest Users. 
	- This option is disabled and thus we are unable to enforce MFA only for Guest Users. 
	- Expectations: Since MFA is enabled, the option to enforce MFA for Guest Users should also be enabled. 
		
**2) Unable to complete MFA Process. Getting an Empty Response and an error "Failed to fetch"**
	- Login to System Console and ensure MFA is enabled. 
	- Set Enforce MFA to true and save.
	- User will be redirected to Configure MFA page. Try to scan the QR Code and submit the token. 
	- The PUT request sent to `/api/v4/users/<USER_ID>/mfa` endpoint receives an EMPTY_RESPONSE and an error `Failed to fetch` is displayed in the UI.

**Comment by jespino** [comment]
14. @michaelgamble what do you thing should be the behavior there? ellipsis? Is the same behavior for users?

**Comment by jespino** [comment]
15. Fixed

**Comment by jespino** [comment]
The 1 from @srkgupta about MFA I think is the expected behavior, can you confirm that @wiersgallak?

**Comment by jespino** [comment]
About @ogi-m bugs:

1. Fixed
2. Fixed
3. Fixed
4. Still need to check that
5. Fixed
6. Fixed
7. Wont fix
8. Fixed
9. Fixed
10. Fixed
11. Fixed
12. Fixed
13. Pending
14. Pending
15. Fixed

**Comment by srkgupta** [comment]
Hi @jespino 

Regarding the issue `Unable to Enforce MFA only for Guest Users`, if you look into this [design](https://projects.invisionapp.com/share/7POW2GOGFVT#/screens/369029392), a message is displayed under the field 'Enforce Multi-factor authentication' indicating the reason why it's disabled, i.e. `MFA is currently not enabled`. Hence it is evident that if MFA is enabled, the option to Enforce MFA for Guest Users should be enabled as well. 

Anyways will wait for a confirmation from @wiersgallak

**Comment by jespino** [comment]
@srkgupta Domain restrictions fixed here: https://github.com/mattermost/mattermost-server/pull/11660

**Comment by jespino** [comment]
@srkgupta From your first numered list:

1 and 2 fixed here: https://github.com/mattermost/mattermost-webapp/pull/3158
3. Was a decision made after the design, support demote to easy the initial steps of adoption of the feature.
4 and 5 are part of another feature that the cloud team is working on, after that feature is implemented this will be implemented. So "wont fix" for now.

**Comment by wiersgallak** [comment]
> The 1 from @srkgupta about MFA I think is the expected behavior, can you confirm that @wiersgallak?

@srkgupta & @jespino Yes, our spec was for when MFA enable is true and MFA enforce is true for regular users, then the option to Enforce MFA for guests would be available.  MFA should be optional  when enable MFA is true and MFA enforce is false.  We will need to use the updated message when MFA enabled = false under Authentication > MFA: 

>Regarding the issue Unable to Enforce MFA only for Guest Users, if you look into this design, a message is displayed under the field 'Enforce Multi-factor authentication' indicating the reason why it's disabled, i.e. MFA is currently not enabled. Hence it is evident that if MFA is enabled, the option to Enforce MFA for Guest Users should be enabled as well.

**Comment by srkgupta** [comment]
Thanks for the update @jespino and thanks for the confirmation @wiersgallak

**Comment by srkgupta** [comment]
Hi @jespino 

Few more issues observed:

**11) Guest user is not able to initiate a Direct Message chat with anyone in the Channel**
- Login as a Guest User and click on + icon next to Direct Messages
- Select a user to initiate a Direct Message chat
- Guest user is not allowed to initiate a chat and an error is thrown in the JS console as shown in this screenshot

DirectMessagesChat

**12) Searching through Channels shows infinite loading and a JS error**
- Login as a Guest user and use the shortcut key to Switch Channels
- Type in something like Guest and check the console 
- As shown in this screenshot, an error is shown and the UI shows an infinite loading. 

SearchChannel2
SearchChannel

**13) Updating the Channel Header does not reflects for the Guest User**
- As shown in these screenshots, a SysAdmin had updated the Channel Header
- However, the Guest user was not able to see the Channel Header (as shown in this screenshot)
- Only after sometime (not sure when), the channel header started appearing
Guest_ChannelHeader

**Comment by srkgupta** [comment]
Hi @jespino, 

The build is broken and we are completely blocked because of it. Can you please check and fix this issue ASAP. 

https://build.mattermost.com/blue/organizations/jenkins/mp%2Fmattermost-server-pr-new/detail/PR-11428/29/pipeline

**Comment by jespino** [comment]
@srkgupta Sure, I'll check it now, it only needs just a merge from master and fixing certain things

**Comment by cpanato** [comment]
I will look why failed

**Comment by cpanato** [comment]
im debugging

**Comment by cpanato** [comment]
@jespino @srkgupta fixed

**Comment by jespino** [comment]
@cpanato Thanks!

**Comment by ogi-m** [comment]
**Guests labelled as "Members" on User Configuration**
- Log in as an admin and go to System Console > User Management
- Click on a guest user
- Note the "Role"
Observed: Role is listed as "Member"
Expected: Role should be "Guest"

**Comment by jespino** [comment]
Updates:

@ogi-m:

1. Fixed
2. Fixed
3. Fixed
4. Fixed
5. Fixed
6. Fixed
7. Wont fix
8. Fixed
9. Fixed
10. Fixed
11. Fixed
12. Fixed
13. Fixed
14. Waiting for @michaelgamble response
15. Fixed
16. Pending

@srkgupta:

1. Fixed
2. Fixed
3. Working as expected
4. Wont fix for now
5. Wont fix for now
6. Wont fix for now
7. Working as expected
8. I think is not related to the PR
9. Decided to defer it until phase2
10. Fixed
11. Can not reproduce
12. Pending
13. Fixed
14. Fixed
15. Can not reproduce

**Comment by michaelgamble** [comment]
@jespino @ogi-m 
14. Hmm this is a tough one.. I think allow the left column to expand past its 50% width shrinking the right column proportionally.. causing the contents of the right column to line wrap as much as needed.. I think its really important to preserve seeing the email in full in this summary screen... also keep in mind the column headers need to be adjusting in width along with it.

**Comment by cpanato** [comment]
@jespino @srkgupta you dont need to remove and re-add the label to get a new test server, every new push will update the test server, this avoids errors as well

**Comment by srkgupta** [comment]
Hi @cpanato 

Sure. We had issues with the previous build. The latest changes were not picked up from the mattermost-webapp repository and hence wanted to give this a try, i.e. to remove the label, destroy the instance and initiate the instance again.

**Comment by cpanato** [comment]
this is another issue, which i will look later

**Comment by srkgupta** [comment]
Ok Thanks for letting me know. Will take care going forward.

**Comment by cpanato** [comment]
it get the branch but maybe an old one
```
+ git checkout PR-11428

error: pathspec 'PR-11428' did not match any file(s) known to git.

+ git checkout feature-guest-accounts

Previous HEAD position was f0467fa... Fix failing E2E test due to recent DOM structure change (#3182)

Switched to a new branch 'feature-guest-accounts'

Branch feature-guest-accounts set up to track remote branch feature-guest-accounts from origin.

+ rm -rf ./dist

+ git rev-parse HEAD

+ export WEBAPP_GIT_COMMIT=ecd76edf876324fdb157422a06e1fdad90eeb4bf

+ echo Webapp Commit: ecd76edf876324fdb157422a06e1fdad90eeb4bf

Webapp Commit: ecd76edf876324fdb157422a06e1fdad90eeb4bf

+ curl -f -o ./dist.tar.gz https://releases.mattermost.com/mattermost-webapp/commit/ecd76edf876324fdb157422a06e1fdad90eeb4bf/mattermost-webapp.tar.gz

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current

                                 Dload  Upload   Total   Spent    Left  Speed


  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  8 26.0M    8 2175k    0     0  10.3M      0  0:00:02 --:--:--  0:00:02 10.3M
100 26.0M  100 26.0M    0     0  31.5M      0 --:--:-- --:--:-- --:--:-- 31.5M

+ mkdir ./dist

+ tar -xvf ./dist.tar.gz -C ./dist --strip-components=1

client/16.c7a3424b9fba68aede1f.js

client/0.ffa8cb46a7f5b15518c5.js

client/12.bc85a8711b21da40f9e7.js

client/20.4f89ed4770e9ad7d98fd.js.map

client/icon_60x60.png

client/8.76f2391c2b3e9ae6c673.js.map

client/15.6911e12e5ccf3fb05be4.js.map

client/33.26dc9c4da4780cbb70a7.js.map

client/icon_57x57.png

client/21.44752c8d8840ee94d162.js.map

client/images/

client/images/warning.png

client/images/img_trans.gif

client/images/appIcons.png

client/images/logo-email.png

client/images/favicon/
....
```

**Comment by cpanato** [comment]
@jespino @srkgupta if you want to the server get the latest webapp please rebase this branch or ammend a commit or whatever, so it will build again

**Comment by srkgupta** [comment]
@jespino looks like the latest changes are still not deployed correctly. Can you please check. I am still able to see some of the issues which were fixed, like Guest Access (Beta) is not shown in the System console, etc.

**Comment by cpanato** [comment]
need to do what i mention above, when the job ran was another sha in the webapp part, to get the most up-to-date need to push a new commit here to trigger the build again and then the server update

**Comment by cpanato** [comment]
now is building with a62247336fefa0f6d1598895063a036fd3e6500d from webapp which is the latest

**Comment by srkgupta** [comment]
Just an update regarding the fixed issues:

Issues 1, 2 ,10,13 and 14 reported by me were fixed and are working fine now. 

New issues observed:
16) Guest User is allowed to delete a post in a direct message even when this option is restricted in the Permission Schemes
- Login as a system admin and navigate to Permission Schemes and ensure Guest User does not has option to Delete Posts
- Login as a Guest User and initiate a direct message chat with one person. 
- Send a message and check if that message post has option to Delete. Try to delete and the message gets deleted
Expectations: Delete option should not be shown for posts by Guest users since this option is restricted as per Permission schemes. 

17) Guest users are still forced to setup MFA when Enforce MFA is turned on for Users and Enforce MFA is disabled for Guest Users
- Login as System Admin and navigate to MFA. Ensure MFA is turned on and is enforced for all users. 
- Navigate to Guest Access page and ensure MFA is not enforced 
- Invite a Guest User and login as Guest User
- Guest User is forced to Setup MFA immediately. 
Expectations: Since the option to enforce MFA for Guest Users is disabled, Guest Users should not be enforced to setup MFA.

**Comment by ogi-m** [comment]
@jespino Regarding my issue #12 - message text is updated but there's still no option to add existing members to team/channels

**Comment by jespino** [comment]
@ogi-m the 12 is deferred to the next phase

**Comment by ogi-m** [comment]
**16. Custom message missing styling**
- As a system admin invite a guest user to join a team and include a custom message
- Go to inbox and open the invite email

**Observed**: Custom message in the email is shown immediately below the "Join Team" button and is missing the styling as seen in the [design](https://projects.invisionapp.com/d/main/default/?origin=v7#/console/17561763/364046309/preview)
**Expected**: Custom message styling matches the design



**17. Guest is taken to Town Square on first log-in**
- As a system admin invite a guest user to join a team and don't add the guest to "Town Square"
- Go to inbox and open the invite link
- Complete the signup process and log in

**Observed**: Guest is taken to Town Square which lands them on "Channel Not Found" page if Town Square isn't included in the list of channels they were added to
**Expected**: Guest is taken to the first channel based on alphabetical order on first log-in


**18. Wrong message and not logged-out when removed from channels**
- Log in as a guest
- In a different browser log in as a system admin and remove the guest from all channels or from the team
- Go back to the browser where the guest user is logged-in

**Observed**: _"You don't belongs to any team and as a guest you can't join any team by yourself, you need to be invited or added by someone else."_
[Expected](https://projects.invisionapp.com/d/main/default/?origin=v7#/console/17561763/364046315/preview): _"Your guest account has no channels assigned. Please contact an administrator."_



**19. System message doesn't specify that a user was added as a guest**
- As a system admin add a guest user to a channel
- Observe the system message in the channel

**Observed**: 'guest-username' added to the channel by 'admin-username'
**Expected** ([per design](https://projects.invisionapp.com/d/main/default/?origin=v7#/console/17561763/364046318/preview)): 'guest-username' added to the channel as a guest by 'admin-username'


**20. "Channel not found" page has "Back to Town Square" link**
- As guest user try to open a channel you previously had access to ( directly via URL for example)

**Observed**: Page shows " Back to Town Square" link.
In case the guest isn't added to Town Square, this link will actually take the guest to a channel that they do have access to. 
**Expected**: The wording for this link should probably be different since Town Square isn't a channel that Guests necessarily belong to? @michaelgamble @jespino

File: store/sqlstore/tokens_store.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ func NewSqlTokenStore(sqlStore SqlStore) store.TokenStore {
 		table := db.AddTableWithName(model.Token{}, "Tokens").SetKeys(false, "Token")
 		table.ColMap("Token").SetMaxSize(64)
 		table.ColMap("Type").SetMaxSize(64)
-		table.ColMap("Extra").SetMaxSize(128)
+		table.ColMap("Extra").SetMaxSize(2048)
```
Comment (grundleborg) [inline store/sqlstore/tokens_store.go]: What happens if this table already exists?

Comment (grundleborg) [reply store/sqlstore/tokens_store.go]: As per discussion on MM, this will be fixed with a migration post-merge.

**Comment by jespino** [comment]
@ogi-m @srkgupta I copied all your bugs here to jira, please, any new bug create it on jira to track them better

**Pull request merged**